temporal-testing/src/main/java/io/temporal/internal/testservice/TestWorkflowMutableStateImpl.java

```diff
@@ -2115,6 +2126,209 @@ public void completeQuery(QueryId queryId, RespondQueryTaskCompletedRequest comp
+            .filter(sm -> !sm.getState().isTerminal())
+            .map(this::constructPendingActivityInfo)
+            .collect(Collectors.toList());
+
+    List<PendingChildExecutionInfo> pendingChildren =
+        this.childWorkflows.values().stream()
+            .filter(sm -> !sm.getState().isTerminal())
+            .map(this::constructPendingChildExecutionInfo)
+            .collect(Collectors.toList());
+
+    return DescribeWorkflowExecutionResponse.newBuilder()
+        .setExecutionConfig(executionConfig)
+        .setWorkflowExecutionInfo(executionInfo)
+        .addAllPendingActivities(pendingActivities)
+        .addAllPendingChildren(pendingChildren)
+        .build();
+  }
+
+  private PendingChildExecutionInfo constructPendingChildExecutionInfo(
+      StateMachine<ChildWorkflowData> sm) {
+    ChildWorkflowData data = sm.getData();
+    return PendingChildExecutionInfo.newBuilder()
+        .setWorkflowId(data.execution.getWorkflowId())
+        .setRunId(data.execution.getRunId())
+        .setWorkflowTypeName(data.initiatedEvent.getWorkflowType().getName())
+        .setInitiatedId(data.initiatedEventId)
+        .setParentClosePolicy(data.initiatedEvent.getParentClosePolicy())
+        .build();
+  }
+
+  private PendingActivityInfo constructPendingActivityInfo(StateMachine<ActivityTaskData> sm) {
+    /*
+     * Working on this code? First, go read StateMachines.scheduleActivityTask to learn which
+     * information it puts in pendingActivity.scheduledEvent and which information it puts
+     * in pendingActivity.activityTask. The ActivityTaskData itself also has some of what we need.
+     */
+    ActivityTaskData pendingActivity = sm.getData();
+    PollActivityTaskQueueResponse.Builder task = pendingActivity.activityTask.getTask();
+    ActivityTaskScheduledEventAttributes scheduledEvent = pendingActivity.scheduledEvent;
+
+    State state = sm.getState();
+    PendingActivityInfo.Builder builder =
+        PendingActivityInfo.newBuilder()
+            .setActivityId(scheduledEvent.getActivityId())
+            .setActivityType(scheduledEvent.getActivityType())
+            .setState(computeActivityState(state, pendingActivity));
+
+    // In golang, we set one but never both of these fields, depending on the activity state
+    if (builder.getState() == PendingActivityState.PENDING_ACTIVITY_STATE_SCHEDULED) {
+      builder.setScheduledTime(task.getScheduledTime());
+    } else {
+      builder.setLastStartedTime(task.getStartedTime());
+      builder.setLastHeartbeatTime(task.getStartedTime());
+    }
+
+    if (pendingActivity.lastHeartbeatTime > 0) {
+      // This may overwrite the heartbeat time we just set - that's fine
+      builder.setLastHeartbeatTime(timestampFromMillis(pendingActivity.lastHeartbeatTime));
+
+      if (pendingActivity.heartbeatDetails != null) {
+        builder.setHeartbeatDetails(pendingActivity.heartbeatDetails);
+      }
+    }
+
+    if (scheduledEvent.hasRetryPolicy()) {
```

Comment:
`scheduledEvent. retryPolicy ` and `pendingActivity.retryState.retryPolicy ` are the same thing. If you replace the first one with the second one here - the code will become cleaner and easier to read.
Also after that, this method can be nicely split into two more granular ones: one is publishing from `scheduledEvent` only, another is publishing from `pendingActivity` only.